RAT/controls.py

```diff
@@ -0,0 +1,212 @@
+import tabulate
+from typing import Union, Any
+from pydantic import BaseModel, Field, field_validator, model_validator
+from RAT.utils.enums import ParallelOptions, Procedures, DisplayOptions, BoundHandlingOptions, StrategyOptions
+
+
+class BaseProcedure(BaseModel, validate_assignment = True, extra = 'forbid'):
+    """
+    Defines the base class with properties used in all five procedures.
+    """
+    parallel: ParallelOptions = ParallelOptions.Single
+    calcSldDuringFit: bool = False
+    resamPars: list[Union[int, float]] = Field([0.9, 50], min_length = 2, max_length = 2)
+    display: DisplayOptions = DisplayOptions.Iter
+
+    @field_validator("resamPars")
+    def check_resamPars(cls, resamPars):
+        if not 0 < resamPars[0] < 1:
+            raise ValueError('resamPars[0] must be between 0 and 1')
+        if not 0 <= resamPars[1]:
+            raise ValueError('resamPars[1] must be greater than or equal to 0')
+        return resamPars
+
+    @classmethod
+    def verify_inputs(cls,
+                      data: dict[str, Any],
+                      expected_properties: list[str],
+                      msg: str) -> None:
+        if isinstance(data, dict) and data != {}:
+            if not all(x in expected_properties for x in data.keys()):
+                raise ValueError(msg)
+
+
+class Calculate(BaseProcedure, validate_assignment = True, extra = 'forbid'):
+    """
+    Defines the class for the calculate procedure.
+    """
+    procedure: Procedures = Field(Procedures.Calculate, frozen = True)
+     
+    @model_validator(mode = 'before')
+    @classmethod
```

Comment:
I added this because I wanted to change the error message and make it more descriptive so that it prints out what attributes can be set for the class. But I can remove it. I just have to undo one commit.

Reply:
Ok, in that case the approach I would recommend is to use a try-except statement and catch the exception, and then compose a custom error - check out `custom_errors.py` for an example.

Reply:
I did it with the try except in Controls Class but it looked messy and that is why I was trying to do it this way. I have removed the model_validation and changed the test. I think it is okay now.

Reply:
We might be able to tidy it up a little (and I think there might be some advanced pydantic tricks I can use as well), but this is the approach that I strongly recommend.